website/archiver/model.py

```diff
@@ -0,0 +1,151 @@
+
+    def __repr__(self):
+        return "<{0}>({1}:{2})".format(self.__class__.__name__, self.name, self.status)
+
+class ArchiveJob(StoredObject):
+
+    _id = fields.StringField(
+        primary=True,
+        default=lambda: str(ObjectId())
+    )
+
+    done = fields.BooleanField(default=False)
+    sent = fields.BooleanField(default=False)
+    status = fields.StringField(default=ARCHIVER_INITIATED)
+    datetime_initiated = fields.DateTimeField(default=datetime.datetime.utcnow)
+
+    dst_node = fields.ForeignField('node', backref='active')
+    src_node = fields.ForeignField('node')
+    initiator = fields.ForeignField('user')
+
+    target_addons = fields.ForeignField('archivetarget', list=True)
+
+    # This field is used for stashing embargo URLs while still in the app context
+    meta = fields.DictionaryField()
+
+    @property
+    def children(self):
+        return [node.archive_job for node in self.dst_node.nodes if node.primary]
+
+    @property
+    def parent(self):
+        parent_node = self.dst_node.parent_node
+        return parent_node.archive_job if parent_node else None
+
+    @property
+    def success(self):
+        return self.status == ARCHIVER_SUCCESS
+
+    def info(self):
+        return self.src_node, self.dst_node, self.initiator
+
+    def target_info(self):
+        return [
+            {
+                'name': target.name,
+                'status': target.status,
+                'stat_result': target.stat_result,
+                'errors': target.errors
+            }
+            for target in self.target_addons
+        ]
+
+    def _archive_node_finished(self):
+        return not any([
+            target for target in self.target_addons
+            if target.status not in (ARCHIVER_SUCCESS, ARCHIVER_FAILURE)
+        ])
+
+    def archive_tree_finished(self):
+        if self._archive_node_finished():
+            return len(
+                [
+                    ret for ret in [
+                        child.archive_tree_finished()
+                        for child in self.children
+                    ] if ret]
+            ) if len(self.children) else True
+        return False
+
+    def _fail_above(self):
+        parent = self.parent
+        if parent:
+            parent.status = ARCHIVER_FAILURE
+            parent.save()
+
+    def _post_update_target(self):
+        if self.status == ARCHIVER_FAILURE:
+            return
+        if self._archive_node_finished():
+            self.done = True
+            if not ARCHIVER_FAILURE_STATUSES.isdisjoint(
+                [target.status for target in self.target_addons]
+            ):
+                self.status = ARCHIVER_FAILURE
+                self._fail_above()
+            else:
+                self.status = ARCHIVER_SUCCESS
+            self.save()
+
+    def get_target(self, addon_short_name):
+        try:
+            return [addon for addon in self.target_addons if addon.name == addon_short_name][0]
```

Comment:
A generator or for loop would be slightly more efficient here, doesn't matter very much.